### PR TITLE
Feature/merge quickstarts

### PR DESCRIPTION
**Pull request by joshmarsh**
## Proposed changes

Consolidating how-to's and quickstarts into one document as per https://github.com/Optum/dce/issues/159

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (changes to code, which do not change application behavior)

## Checklist

- [x] I have filled out this PR template
- [x] I have read the [CONTRIBUTING](https://github.com/Optum/dce/blob/master/docs/CONTRIBUTING.md) doc
- [ ] I have added automated tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (`README.md`, inline comments, etc.)
- [x] I have updated the `CHANGELOG.md` under a `## next` release, with a short summary of my changes

### CODE REVIEW

File: docs/howto.md [hunk 1]
```diff
@@ -83,27 +373,214 @@ The child account must have an administrative IAM Role with a trust relationship
 }
 ```
 
-Add the account to the pool using the DCE API:
+Add the account to the pool using the DCE API
+
+**Request**
+
+`POST ${api_url}/accounts`
```
Comment (joshmarsh) [inline docs/howto.md]: It would be great if we could somehow provide [swagger ui links](https://petstore.swagger.io/?_ga=2.251168133.206317326.1577721714-338265257.1569482972#/pet/uploadFile) rather than manually updating these every time something changes.

File: docs/api-auth.md [hunk 2]
```diff
@@ -7,13 +7,15 @@ There are two ways to authenticate against the DCE APIs:
 
 ## Using Cognito
 
-Cognito is used to allow an admin to add users to DCE.  This can be done by using the Cognito User Pool.  Additionally any IdP supported by Cognito User Pools can also be used.
+Cognito is used to allow an admin to add users to DCE.  This can be done by using the Cognito User Pool.  This section will walk through setting this
```
Comment (eschwartz) [inline docs/api-auth.md]: > Cognito is used to allow an admin to add users to DCE.

That not quite right. Cognito is used to authenticate and authorize DCE API users. 
Maybe the terminology is just getting confusing here, as we're talking about _admins_ in a couple different contexts (ie, _System admin_ can use Cognito to manage _API users_. But there's no concept of _API Admins_ managing _DCE users_)

Comment (joshmarsh) [reply docs/api-auth.md]: Good catch, I'll rewrite this section.

File: docs/api-auth.md [hunk 3]
```diff
@@ -25,12 +27,94 @@ There are three different ways a user is considered an admin:
...
+    []
+    ```
+
 ## Using IAM Credentials
 
 
 The DCE API accepts authentication via IAM credentials using [SigV4 signed requests](https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html).
 
-Any requests made via IAM Credentials will be treated as an [admin role](#admins). 
+Any requests made via sufficiently permissioned IAM Credentials will be treated as an [admin role](#admins).
```
Comment (eschwartz) [inline docs/api-auth.md]: Maybe some clarification here? 
Any IAM creds are treated as admin, if they make their way to the Lambda. 
Maybe what you're getting at is that the IAM creds need a policy with access to our API gateway endpoints. I think we even have a resource in TF for this, that we could document here

File: docs/api-auth.md [hunk 4]
```diff
@@ -25,12 +27,94 @@ There are three different ways a user is considered an admin:
 
 Users (by default) are given access to the leasing and usage APIs.  This is done so they can request their own lease and look at the usage of their leases.  Any appropriately authenticated user in Cognito will automatically fall into the `Users` role.
 
+### Configuring Cognito for Usage with the DCE CLI
```
Comment (eschwartz) [inline docs/api-auth.md]: Scrap _with the DCE CLI_ from this section heading.

We use the DCE CLI to demonstrate how DCE works, but this section is really about _Configuring DCE Users in Cognito_

File: docs/api-auth.md [hunk 5]
```diff
@@ -40,7 +124,9 @@ The process for signing requests with SigV4 is somewhat involved, but luckily th
 
 AWS also provides [examples for a number of languages in their docs](https://docs.aws.amazon.com/general/latest/gr/signature-v4-examples.html).
 
+### Using IAM Credentials with the DCE CLI
 
+If there is no api auth token present in the DCE configuration file, the CLI will look for AWS credentials in `$HOME/.aws/credentials` (Linux/OSX) or `%USERPROFILE%\.aws\credentials` (Windows).
```
Comment (eschwartz) [inline docs/api-auth.md]: The CLI will also look for AWS env vars, for creds.

**Comment by eschwartz** [comment]
See PR for fixing `make documentation` command: #193 
(to make this easier for folks to review)

**Comment by joshmarsh** [comment]
> I think we should keep a quickstart doc.
> The How To guide is like a recipe book, with more advanced customizations for DCE.
> As a new user to the system, I want something short and sweet that gets me in up and running with as little effort as possible.

I don't think the [current quickstart](https://dce.readthedocs.io/en/latest/quickstart.html) is much quicker than following the [how to](https://github.com/Optum/dce/pull/173/files?short_path=63cc851#diff-63cc851060ce7b11dd6db47796545ca9) sections on using the cli or api. Now that I've moved the Cognito stuff into api-auth.md document, the CLI section of the how-to.md pretty much reflects the bare minimum information you need to both deploy  dce and do basic CRUD operations on accounts/ leases.

Keeping in mind that any additional documentation we write will need to be updated as things change, I think it makes sense for us to focus on maintaining DRY, concise, and accurate documentation while we focus on working through some of our open github issues.

**Comment by eschwartz** [comment]
I'm taking another 👀

**Comment by eschwartz** [comment]
@joshmarsh or @nathanagood any idea why some of the code snippets have red boxes around each line?
image
image

File: README.md [hunk 6]
```diff
@@ -10,13 +10,13 @@ At the end of the lease, or if the lease's budget is reached, the account is wip
 
 ## Getting Started & Documentation
 
-Deploy your own Disposable Cloud Environment by following our [quick start guide](./docs/quickstart.md), available on our documentation website:
+Deploy your own Disposable Cloud Environment by following the [how-to guide](./docs/howto.md), also available on our documentation website:
```
Comment (eschwartz) [inline README.md]: Let's keep this pointing at the quick start guide

File: docs/api-auth.md [hunk 7]
```diff
@@ -2,18 +2,20 @@
...
 
-## Using Cognito
+## Using AWS Cognito
 
-Cognito is used to allow an admin to add users to DCE.  This can be done by using the Cognito User Pool.  Additionally any IdP supported by Cognito User Pools can also be used.
+AWS Cognito is used to authenticate and authorize DCE users. This section will walk through setting this
+up in the AWS web console, but note that all of these operations can be automated using the AWS CLI or SDKs. While
+this example uses Cognito User Pools to create and manage users, you may also [integrate Cognito with your own IdP](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-identity-provider.html).
 
 ### Roles
```
Comment (eschwartz) [inline docs/api-auth.md]: I think this should be a second-level heading -- it's not specific to Cognito

File: docs/api-auth.md [hunk 8]
```diff
@@ -25,30 +27,123 @@ There are three different ways a user is considered an admin:
...
-The Terraform module in the repo provides an IAM policy with appropriate permissions for executing DCE API requests. The policy name and ARN are available as Terraform outputs.
+1. Users are not authorized to list child accounts in the accounts pool. Type `dce accounts list` to verify that you get a permissions error when trying to
+view information you do not have access to.
 
+    ```
+    dce accounts list
+    err:  [GET /accounts][403] getAccountsForbidden
+    ```
+
+1. You will need to be authenticated as an admin before continuing to the next section. Type `dce auth` to log in as a different user. Sign out, then enter the username
```
Comment (eschwartz) [inline docs/api-auth.md]: This is a little confusing to me to have this in here.
I get you're trying to demonstrate the different auth levels, but it makes it seem like it required to switch roles as part of setting up cognito. I would cut it.

File: docs/api-auth.md [hunk 9]
```diff
@@ -25,30 +27,123 @@ There are three different ways a user is considered an admin:
...
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "execute-api:Invoke"
+      ],
+      "Resource": [
+        "${aws_api_gateway_rest_api.gateway_api.execution_arn}/*"
```
Comment (eschwartz) [inline docs/api-auth.md]: maybe confusing that we're including terraform interpolation syntax inside an IAM policy?

I would either do something like:

```json
 "Resource": [
        "<DCE API Gateway ARN>"
  ]
```
or, just put  a wildcard, and let them figure it out if they want it more restrictive

File: docs/api-auth.md [hunk 10]
```diff
@@ -25,30 +27,123 @@ There are three different ways a user is considered an admin:
...
+}
+JSON
+}
 ```
-terraform output api_access_policy_name
-terraform output api_access_policy_arn
-```
\ No newline at end of file
+
+This policy is accessible via the api_access_policy_name & api_access_policy_arn terraform outputs.
```
Comment (eschwartz) [inline docs/api-auth.md]: Not necessarily explanatory how to access these. Maybe link to /terraform.html#accessing-terraform-outputs

File: docs/api-auth.md [hunk 11]
```diff
@@ -25,30 +27,123 @@ There are three different ways a user is considered an admin:
...
+
+Any requests made with IAM credentials that have sufficient permissions to invoke the DCE API, but which are not associated with a Congito User Pool User, will be treated as an `admin role <#admins>`_.
+
+The process for signing requests with SigV4 is somewhat involved, but luckily there are a number of tools to make this easier. For example:
```
Comment (eschwartz) [inline docs/api-auth.md]: I think this page is getting a little confused, as it's mixing instructions for authenticating directly against the API together with directions for authenticating with DCE CLI.

If we're just talking DCE CLI, we don't need to talk about sig v4. But we _do_ need a section on where DCE CLI gets credentials from, and in what order. 
I guess I'd like to see this page re-org'd a bit, like

1. Configuring API User with AWS IAM
1. Configuring API Users with Cognito
1. Authenticating requests with the DCE CLI
1. Authenticating requests to the DCE API
1. Authorization Roles

what do you think?

File: docs/changelog_link.rst [hunk 12]
```diff
@@ -1,3 +0,0 @@
-Changes
```
Comment (eschwartz) [inline docs/changelog_link.rst]: I guess I'm not super attached to it, but what's the reasoning behind removing the changelog from the docs?

File: docs/concepts.md [hunk 13]
```diff
@@ -3,7 +3,7 @@
 ## DCE
 
 _Disposable Cloud Environment (DCE)_ provide temporary, limited access to Amazon Web 
-Services (AWS) accounts. Administrators can configure this limited access to expire based on time or budget. When the access expires, DCE destroys all of the resources in the account and returns the [account to the [account pool](#account-pool)
+Services (AWS) accounts. Administrators can configure this limited access to expire based on time or budget. When the access expires, DCE destroys all of the resources in the account and returns the account to the [account pool](#account-pool).
```
Comment (eschwartz) [inline docs/concepts.md]: This link isn't rendering
image

File: docs/concepts.md [hunk 14]
```diff
@@ -74,7 +74,7 @@ the configurable default is unset, DCE uses a period of seven (7) days.
 A lease that is _over budget_ has exceeded the budget amount set
 by the `budgetAmount` field of the lease.
 
-Each lease has a configurable _[budget](#budget)_. DCE periodically
+Each lease has a configurable [budget](#budget). DCE periodically
```
Comment (eschwartz) [inline docs/concepts.md]: Is this PR attempting to fix all the links? this one isn't rendering, and I'm still seeing a bunch of errors from Sphinx.
 Maybe we should put this effort in another PR?

File: docs/home.md [hunk 15]
```diff
@@ -34,7 +34,7 @@ for other people to lease.
 
 ## Getting started
 
-To get started using DCE, see the [Quickstart](quickstart.md).
+To get started using DCE, see the [how-to guide](howto.md).
```
Comment (eschwartz) [inline docs/home.md]: Let's keep this pointing at the quick start

File: docs/quickstart.md [hunk 16]
```diff
@@ -1,246 +1,67 @@
...
-```
+Deploy DCE and lease an account quickly using the DCE CLI.
+
+1. Download the appropriate executable for your OS from the [latest release](https://github.com/Optum/dce-cli/releases/latest). e.g. for mac, you should download dce_darwin_amd64.zip
+
+1. Unzip and move the executable to a directory on your PATH, e.g.
+
+    ```
+    # Download the zip file
+    wget https://github.com/Optum/dce-cli/releases/download/<VERSION>/dce_darwin_amd64.zip
```
Comment (eschwartz) [inline docs/quickstart.md]: I think you can just replace `<VERSION>` with `latest` here, so this is a working example

File: docs/quickstart.md [hunk 17]
```diff
@@ -1,246 +1,67 @@
...
+1. Type `dce init`. Leave all fields blank for now.
+
+1. Deploy DCE using IAM Credentials that have AdministratorAccess
+    ```
+    export AWS_ACCESS_KEY_ID=XXXXXXXXXX
+    export AWS_SECRET_ACCESS_KEY=XXXXXXXXXXXXXXXXXXXX
+    dce system deploy
+    ```
+
+1. Retrieve the DCE API url from API Gateway in your master account, and add it to the dce config file, e.g.
```
Comment (eschwartz) [inline docs/quickstart.md]: > Retrieve the DCE API url from API Gateway in your master account

How does the user find this?
I think `dce system deploy` will spit this out, no? or is that hard for the user to find?

File: docs/quickstart.md [hunk 18]
```diff
@@ -1,246 +1,67 @@
...
+    ```
+    api:
+        host: abcdefghij.execute-api.us-east-1.amazonaws.com
+        basepath: /api
+    region: us-east-1
+    ```
+
+1. Prepare a second AWS account to be your first "DCE Child Account".
+    - Create an IAM role with `AdministratorAccess` and a trust relationship to your DCE Master Accounts
+    - Create an account alias by clicking the 'customize' link in the IAM dashbodard of the child account. This must not include the terms "prod" or "production".
```
Comment (eschwartz) [inline docs/quickstart.md]: >  clicking the 'customize' link in the IAM dashbodard of the child account

I'm not sure I'd know where to find this.
Maybe stick with CLI commands if we can:
https://docs.aws.amazon.com/cli/latest/reference/iam/create-account-alias.html

File: docs/quickstart.md [hunk 19]
```diff
@@ -1,246 +1,67 @@
...
+
+1. Lease your child account
+    ```
+    dce leases create --budget-amount 100.0 --budget-currency USD --email jane.doe@email.com --principal-id quickstartuser
+    ```
+
+1. Log in to your leased account
+    ```
+    dce leases login --open-browser <lease-id>
+    ```
```
Comment (eschwartz) [inline docs/quickstart.md]: ...."This will open a browser, and bring you to your AWS Account web dashboard"

... and maybe worth just an extra step, to show getting CLI creds (w/o `--open-browser` flag)

**Comment by joshmarsh** [comment]
> @joshmarsh or @nathanagood any idea why some of the code snippets have red boxes around each line?
> image
> image

I

**Comment by joshmarsh** [comment]
> @joshmarsh or @nathanagood any idea why some of the code snippets have red boxes around each line?
> image
> image

Not sure. I don't see anything in the conf.py that looks like it would cause that. @nathanagood ?

File: docs/changelog_link.rst (hunk 12 above)
Comment (joshmarsh) [reply docs/changelog_link.rst]: We'd discussed it not being needed in readthedocs, so I left it out after initially changing from m2r -> recommonmark

Comment (marinatedpork) [reply docs/changelog_link.rst]: I am on board with removing the change log, but make sure the API docs still work in readthedocs when changing any of the rendering engines

File: docs/quickstart.md (hunk 17 above)
Comment (marinatedpork) [reply docs/quickstart.md]: We could go a step further and populate .dce/config with the TF outputs on deploy complete.

Comment (joshmarsh) [reply docs/quickstart.md]: It's currently too hard to find. I think this is something we want to automate in the near future as well.

File: docs/quickstart.md (hunk 18 above)
Comment (marinatedpork) [reply docs/quickstart.md]: I think the UI directions are good in addition to the link to the docs. Sentence can be rewritten as:

> Create an alias by going to the IAM Dashboard and clicking the "customize" link or by [using the AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/iam/create-account-alias.html).
